### DISCUSSION TITLE
Saving the data with subject ID in file name

### DISCUSSION THREAD

**Discussion by becky-gilbert** [jspsych/jsPsych#1699] [Show and tell]
Here's how to generate a random ID for each participant, save it to the data, and save the data as a local file with this ID in the name.
You can use the [jsPsych.randomization.randomID](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationrandomid) function to generate the ID string, and the [jsPsych.data.addProperties](https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties) function to add this ID to the data for all trials. Then to save a file that includes this ID in the file name, you can create a file name variable by joining text and the ID variable with the + operator.
```js
// generate a random subject ID that contains 8 alphanumeric characters
var rand_subject_id = jsPsych.randomization.randomID(8);
// add the ID to the data for all trials
jsPsych.data.addProperties({ID: rand_subject_id});

var trial = {
  type: 'html-keyboard-response',
  stimulus: 'Hello!'
};

jsPsych.init({
  timeline: [trial],
  on_finish: function () {
    // combine the subject ID with the file extension (and any other text) to dynamically create the file name
    var file_name = 'subject_' + rand_subject_id + '_data.csv';
    // save the data with this file name
    jsPsych.data.get().localSave('csv',file_name);
  }
});
```
Or, if you want to get the ID from a response that was made during the experiment, you can do that by filtering the data to find the data from the ID question using the [`filter` function](https://www.jspsych.org/core_library/jspsych-data/#filter), then getting the raw data values with the [`values` function](https://www.jspsych.org/core_library/jspsych-data/#values), then selecting the first (only) trial in that array, and then finding the value associated with the question called "id" in the "responses" object.
```js
var id_trial = {
  type: 'survey-text',
  questions: [{prompt: 'Please enter your ID.', name: 'id', required: true}],
  data: {task_part: 'id_question'}
};

jsPsych.init({
  timeline: [id_trial],
  on_finish: function () {
    // get the ID that was entered from the jsPsych data
    var id = jsPsych.data.get().filter({task_part: 'id_question'}).values()[0].response.id;
    // add this ID to the data for all trials
    jsPsych.data.addProperties({ID: id});
    // combine the subject ID with the file extension (and any other text) to dynamically create the file name
    var file_name = 'subject_' + id + '_data.csv';
    // save the data with this file name
    jsPsych.data.get().localSave('csv',file_name);
  }
});
```

**Comment by xiaoqilt** [comment]
var id = jsPsych.data.get().filter({task_part: 'id_question'}).values()[0].response.id;

Autofocus processing was blocked because a document already has a focused element.
test1.html:44 
        
       Uncaught TypeError: Cannot read properties of undefined (reading 'id')
    at Object.on_finish (test1.html:44:87)
    at finishExperiment (jspsych.js:815:10)
    at nextTrial (jspsych.js:839:7)
    at window.jsPsych.core.finishTrial (jspsych.js:267:9)
    at HTMLFormElement.<anonymous> (jspsych-survey-text.js:169:15)
